src/modules/fancyzones/lib/ZoneSet.cpp

```diff
@@ -244,30 +244,54 @@ ZoneSet::ZonesFromPoint(POINT pt) const noexcept
 
     if (overlap)
     {
-        size_t smallestIdx = 0;
-        for (size_t i = 1; i < capturedZones.size(); ++i)
+        try
         {
-            RECT rectS;
-            RECT rectI;
-            try
+            auto expand = [&](RECT& rect) {
+                rect.top -= m_config.SensitivityRadius / 2;
+                rect.bottom += m_config.SensitivityRadius / 2;
+                rect.left -= m_config.SensitivityRadius / 2;
+                rect.right += m_config.SensitivityRadius / 2;
+            };
+
+            // Compute the overlapped rectangle.
+            RECT overlap = m_zones.at(capturedZones[0])->GetZoneRect();
+            expand(overlap);
+
+            for (size_t i = 1; i < capturedZones.size(); ++i)
             {
-                rectS = m_zones.at(capturedZones[smallestIdx])->GetZoneRect();
-                rectI = m_zones.at(capturedZones[i])->GetZoneRect();
+                RECT current = m_zones.at(capturedZones[i])->GetZoneRect();
+                expand(current);
+
+                overlap.top = max(overlap.top, current.top);
+                overlap.left = max(overlap.left, current.left);
+                overlap.bottom = min(overlap.bottom, current.bottom);
+                overlap.right = min(overlap.right, current.right);
             }
-            catch (std::out_of_range)
+
+            // Avoid division by zero
+            int width = max(overlap.right - overlap.left, 1);
+            int height = max(overlap.bottom - overlap.top, 1);
+
+            bool verticalSplit = height > width;
+            size_t zoneIndex;
+
+            if (verticalSplit)
             {
-                return {};
+                zoneIndex = (pt.y - overlap.top) * capturedZones.size() / height;
             }
-            int smallestSize = (rectS.bottom - rectS.top) * (rectS.right - rectS.left);
-            int iSize = (rectI.bottom - rectI.top) * (rectI.right - rectI.left);
-
-            if (iSize <= smallestSize)
+            else
             {
-                smallestIdx = i;
+                zoneIndex = (pt.x - overlap.left) * capturedZones.size() / width;
             }
-        }
 
-        capturedZones = { capturedZones[smallestIdx] };
+            zoneIndex = std::clamp(zoneIndex, size_t(0), capturedZones.size() - 1);
+
+            return { capturedZones[zoneIndex] };
+        }
+        catch (std::out_of_range)
+        {
+            return {};
```

Comment:
We should log the error, and possibly default to one of the overlapping zone.

Reply:
This error should never happen and I don't even know why this check was added to the previous version, but sure, logging it is a good idea.